### DISCUSSION TITLE
Calling a classmethod of a TypeVar bound to a Protocol class with return type Self does not resolve

### DISCUSSION THREAD

**Discussion by GuillaumeDesforges** [microsoft/pyright#5661] [General]
I have the snippet

```python
from typing import Any, Callable, ClassVar, Protocol, Self, Type, TypeVar


def test_typevar_bound_protocol_self():
    class SomeProtocol(Protocol):
        some_class_method: ClassVar[Callable[..., Self]]

    T_SomeProtocol = TypeVar(
        "T_SomeProtocol",
        bound=SomeProtocol,
    )

    def some_function(
        some_class: Type[T_SomeProtocol],
    ) -> T_SomeProtocol:
        return some_class.some_class_method()

    class SomeClass:
        @classmethod
        def some_class_method(cls):
            return SomeClass()

    result = some_function(SomeClass)
    assert isinstance(result, SomeClass)


if __name__ == "__main__":
    test_typevar_bound_protocol_self()

```

I can run it with Python, but pyright fails with errors

```
error: Expression of type "Self@SomeProtocol" cannot be assigned to return type "T_SomeProtocol@some_function"
    Type "Self@SomeProtocol" cannot be assigned to type "T_SomeProtocol@some_function"
```

Is it a limitation of pyright?

**Comment by erictraut** [comment]
Pyright is correct in generating an error here — for the reasons indicated in the error message. This isn't a limitation in pyright (mypy generates the same error here). This is simply the Python type system doing what it's designed to do — detect static type violations.

The class method `some_class_method` requires for its first parameter a value that is the class `SomeClass` or a subclass thereof. When function `some_function` invokes the `some_class_method` class method, it fails to provide a value for the `cls` parameter with a compatible type. Instead of providing a value that is guaranteed to be `SomeClass` or a subclass thereof, it's providing a value that conforms to a protocol `SomeProtocol`. That's insufficient for satisfying the type of parameter `cls`, hence a type violation. `SomeClass` is a subtype of `SomeProtocol`, but the inverse is not true. There are an infinite number of types that conform to `SomeProtocol` that are not subtypes of `SomeClass`.

I would suggest a fix to the bug in your code, but I don't know enough about what you're trying to do in this case. Hopefully the above explanation makes sense and you can figure out a proper fix on your own. If not, feel free to provide more context and I'll try to help you out.

**Reply by GuillaumeDesforges** [reply]
Thanks for the detailed reply. It's a good idea to try with mypy, I should have done that.

Unfortunately it looks to me like the error is not when trying to solve the constraints on the first arg of `some_class_method`.
AFAIK the type `Callable[..., Self]` should be assignable to any function that takes some arguments and returns an instance of the class.
Hence I would have expected that, in this case, `SomeClass.some_class_method` would be a class variable of type `Callable[[Type[Self]], Self]`, which should be assignable to `Callable[..., Self]`.

The error seems to come from the assignement of the return type of `some_class.some_class_method()` which appears to be `Self@SomeProtocol` to the return type of `some_function` which is `T_SomeProtocol@some_function`.
Looking at [PEP 673](https://peps.python.org/pep-0673/#use-in-classmethod-signatures), given that `some_class` is of type `Type[T_SomeProtocol]`, I would have expected pyright to infer that its `Self` would be `T_SomeProtocol`.

If I remove the return type of `some_function`

```diff
     def some_function(
         some_class: Type[T_SomeProtocol],
-     ) -> T_SomeProtocol:
+     ):
         return some_class.some_class_method()
```

I have no type error, but pyright lens shows me a type hint `Self` as the return type of `some_function`.

**Reply by GuillaumeDesforges** [reply]
Writing a reply to your answer, I found how to fix my issue:

```python
from typing import Any, Protocol, Self, Type, TypeVar


def test_typevar_bound_protocol_self():
    class SomeProtocol(Protocol):
        @classmethod
        def some_class_method(cls: Type[Self]) -> Self:
            ...

        some_attribute: Any

    T_SomeProtocol = TypeVar(
        "T_SomeProtocol",
        bound=SomeProtocol,
    )

    def some_function(
        some_class: Type[T_SomeProtocol],
    ) -> T_SomeProtocol:
        return some_class.some_class_method()

    class SomeClass:
        @classmethod
        def some_class_method(cls):
            return SomeClass()

        some_attribute = 1

    result = some_function(SomeClass)
    assert isinstance(result, SomeClass)


if __name__ == "__main__":
    test_typevar_bound_protocol_self()
```

This type checks and makes sense to me.
Thanks for the help!
